modules/edge-identity/src/identity.rs

```diff
@@ -128,126 +154,88 @@ decl_module! {
 				IdentityStage::Verified => return Err("Already verified"),
 				IdentityStage::Attested => ()
 			}
 
-			// Check the number of verifications the record has and ensure all are valid
-			let mut valid_verifications: Vec<(T::AccountId, bool)> = [(_sender.clone(), vote)].to_vec();
-			let acct = record.account.clone();
-			if let Some(verifications) = record.verifications {
-				valid_verifications = verifications.clone()
-					.into_iter()
-					.filter(|v| Self::verifiers().contains(&v.0))
-					.collect();
-				valid_verifications.push((_sender.clone(), vote));
-			}
+			ensure!(<timestamp::Module<T>>::get() <= record.expiration_time, "Identity expired");
 
-			let yes_votes: Vec<T::AccountId> = valid_verifications.clone()
-				.into_iter()
-				.filter(|v| v.1 == true)
-				.map(|v| v.0)
-				.collect();
-
-			let no_votes: Vec<T::AccountId> = valid_verifications.clone()
-				.into_iter()
-				.filter(|v| v.1 == false)
-				.map(|v| v.0)
-				.collect();
-
-			// Check if we have gather a supermajority of no votes
-			// TODO: What does it mean to fail a verification? If malicious
-			//		 behavior, we should punish the sending account somehow.
-			if no_votes.len() * 3 >= 2 * Self::verifiers().len() {
-				Self::remove_pending_identity(&identity_hash, true);
-				<FrozenAccounts<T>>::mutate(|froze| froze.push(acct));
-				Self::deposit_event(RawEvent::Failed(identity_hash, record.account.into()));
-			} else {
-				// Check if we have gathered a supermajority of yes votes
-				let mut stage = IdentityStage::Attested;
-				let mut expiration_time = record.expiration_time;
-				let registrar = record.account.clone();
-				if yes_votes.len() * 3 >= 2 * Self::verifiers().len() {
-					stage = IdentityStage::Verified;
-					expiration_time = None;
-				}
-				// Insert updated identity record back into the collection
-				<IdentityOf<T>>::insert(identity_hash, IdentityRecord {
-					stage: stage,
-					expiration_time: expiration_time,
-					verifications: Some(valid_verifications),
-					..record
-				});
+			ensure!(!<VerifiedIdentityBy<T>>::get((identity_hash, _sender.clone())), "Verifier already verified");
 
-				if stage == IdentityStage::Verified {
-					Self::remove_pending_identity(&identity_hash, false);
-					Self::deposit_event(RawEvent::Verify(identity_hash, registrar.into(), yes_votes));
-				}
-			}
+			<VerifiedIdentityBy<T>>::insert((identity_hash, _sender.clone()), true);
+			record.verifications[vote as usize] += 1;
+
+			<IdentityOf<T>>::insert(identity_hash, IdentityRecord {
+				..record
+			});
 
 			Ok(())
 		}
 
-		/// Add metadata to sender's account.
-		// TODO: make all options and only updated provided?
-		// TODO: limit the max length of these user-submitted types?
-		pub fn add_metadata(origin, identity_hash: T::Hash, avatar: Vec<u8>, display_name: Vec<u8>, tagline: Vec<u8>) -> Result {
-			let _sender = ensure_signed(origin)?;
+		pub fn finalize_verification(_origin, identity_hash: T::Hash) -> Result {
```

Comment:
Who do we expect will call `finalize_verification`? The verifier?

Reply:
It can be called by anyone, but verifier would be the most natural candidate. Again, client can enforce such action. After all, if verifier dont want to call it, the identity creator can do it. It would be in his interest to do so.

If you feel like we shouldn't enforce anyone to worry about this function of they don't have to, we can call it automatically inside `verity` if conditions are met (conditions would be: 1) we are having "yes" majority, 2) we are having "no" majority 3) everyone has submitted a vote)

(I realized that, currently `finalize_verification`, can be called at any time, which can lead to identity being removed maliciously. I will add additional checks.)